apps/ejabberd/src/mongoose_api_common.erl

```diff
@@ -0,0 +1,352 @@
+maybe_add_location_header(Result, ResourcePath, Req) when is_float(Result) ->
+    add_location_header(float_to_list(Result), ResourcePath, Req);
+maybe_add_location_header(_R, _R, Req) ->
+    {ok, Req2} = cowboy_req:reply(200, [], Req),
+    Req2.
+
+add_location_header(Result, ResourcePath, Req) ->
+    Path = [ResourcePath, "/", Result],
+    Header = {<<"location">>, Path},
+    {ok, Req2} = cowboy_req:reply(201, [Header], Req),
+    Req2.
+
+
+
+-spec convert_arg(atom(), any()) -> integer() | float() | binary() | string() | {error, bad_type}.
+convert_arg(binary, Binary) when is_binary(Binary) ->
+    Binary;
+convert_arg(integer, Binary) when is_binary(Binary) ->
+    binary_to_integer(Binary);
+convert_arg(integer, Integer) when is_integer(Integer) ->
+    Integer;
+convert_arg(float, Binary) when is_binary(Binary) ->
+    binary_to_float(Binary);
+convert_arg(float, Float) when is_float(Float) ->
+    Float;
+convert_arg(_, _Binary) ->
+    throw({error, bad_type}).
+
+-spec create_params_proplist(list({binary(), binary()})) -> args_applied().
+create_params_proplist(ArgList) ->
+    lists:sort([{to_atom(Arg), Value} || {Arg, Value} <- ArgList]).
+
+-spec category_to_resource(atom() | list()) -> string().
+category_to_resource(Category) when is_atom(Category) ->
+    atom_to_list(Category);
+category_to_resource(Category) when is_list(Category) ->
+    Category.
+
+%% @doc Returns list of allowed methods.
+-spec get_allowed_methods(admin | user) -> list(method()).
+get_allowed_methods(Entity) ->
+    Commands = mongoose_commands:list(Entity),
+    [action_to_method(mongoose_commands:action(Command)) || {_Name, Command} <- Commands].
+
+-spec maybe_add_bindings(mongoose_commands:t(), admin|user) -> string().
+maybe_add_bindings(Command, Entity) ->
+    Action = mongoose_commands:action(Command),
+    Args = mongoose_commands:args(Command),
+    BindAndBody = both_bind_and_body(Action),
+    case BindAndBody of
+        true ->
+            Ids = mongoose_commands:identifiers(Command),
+            Bindings = [El || {Key, _Value} = El <- Args, true =:= proplists:is_defined(Key, Ids)],
+            add_bindings(Bindings, Entity);
+        false ->
+            add_bindings(Args, Entity)
+    end.
+
+-spec both_bind_and_body(mongoose_commands:command_action()) -> boolean().
+both_bind_and_body(update) ->
+    true;
+both_bind_and_body(create) ->
+    true;
+both_bind_and_body(read) ->
+    false;
+both_bind_and_body(delete) ->
+    false.
+
+add_bindings(Args, Entity) ->
+    lists:flatten([add_bind(A, Entity) || A <- Args]).
+
+%% skip "caller" arg for frontend command
+add_bind({caller, _}, user) ->
+    "";
+add_bind({ArgName, _}, _Entity) ->
+    ["/", atom_to_list(ArgName), "/:", atom_to_list(ArgName)];
+add_bind(Other, _) ->
+    throw({error, bad_arg_spec, Other}).
+
+-spec to_atom(binary() | list() | atom()) -> atom().
+to_atom(Bin) when is_binary(Bin) ->
+    list_to_atom(binary_to_list(Bin));
```

Comment:
and imagine there is some nasty client who knows about that and sends thousands of requests with unique data and the new created atoms won't be garbage collected! 
Thanks for pointing it out